src/preprocess/preprocess.cpp

```diff
@@ -0,0 +1,137 @@
+  if (!resolve_relations(query.get_relations(), sch)) {
+    return nullptr;
+  }
+  if (!resolve_attributes(query.get_attributes(), sch)) {
+    return nullptr;
+  }
+  if (!resolve_predicate_leaves(query.get_conditions(), sch)) {
+    return nullptr;
+  }
+  // we assure that parse success here
+  auto ptr = pre_compute_simple_leaves(query.transfer_conditions());
+
+  return std::make_unique<SelectStmt>(sch, query.get_attributes(), std::move(ptr));
+} 
+
+auto Preprocessor::resolve_relations(const std::vector<std::string>& relations, Schema& schema) -> bool {
+  for (auto& rel : relations) {
+    auto sch_arr = ts_manager_.get_relation(rel);
+    if (sch_arr.empty()) {
+      printer_.output_error("unknown relation: " + rel);
+      return false;
+    }
+    for (auto& item : sch_arr) {
+      schema.append_schema(item);
+    }
+  }
+  return true;
+}
+
+auto Preprocessor::resolve_attribute_item(Attribute& attr, const Schema& schema) -> bool {
+  auto& attr_name = attr.get_attribute();
+  auto& rel_name = attr.get_relation();
+  if (rel_name.empty()) {
+    // find relation
+    int rel_num = schema.count_relation(attr_name);
+    if (1 != rel_num) {
+      printer_.output_error(rel_num > 1 ? ("ambiguous attribute: " + attr_name) : ("unknown attribute: " + attr_name));
+      return false;
+    } else {
+      attr.set_relation(schema.get_relation(attr_name));
+    }
+  } else {
+    // check whether exists
+    if (!schema.is_attribute_exists(rel_name, attr_name)) {
+      printer_.output_error("unknown attribute: " + rel_name + "." + attr_name);
+      return false;
+    }
+  }
+  return true;
+}
+
+auto Preprocessor::resolve_attributes(std::vector<Attribute>& attributes, const Schema& schema) -> bool {
+  for (auto& attr : attributes) {
+    if (!resolve_attribute_item(attr, schema)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+auto Preprocessor::resolve_operand_type(Operand& oprd, const Schema& schema) -> ValueType {
+  if (oprd.get_operand_type() == OperandType::COT_ATTR) {
+    auto& oprd_attr = dynamic_cast<OperandAttr&>(oprd);
+    auto& attr = oprd_attr.get_attr();
+    bool flag = resolve_attribute_item(attr, schema);
+    if (!flag) {
+      return ValueType::VT_INVALID;
+    } else {
+      return schema.get_attribute_schema(attr.get_relation(), attr.get_attribute()).type_;
+    }
+  } else {
+    auto& oprd_val = dynamic_cast<OperandVal&>(oprd);
+    return oprd_val.get_value().get_value_type();
+  }
+}
+
+auto Preprocessor::resolve_predicate_leaves(Predicate& conditions, const Schema& schema) -> bool {
+  switch (conditions.get_node_type()) {
+    case PredicateType::CNT_LEAF: {
+      auto& leaf = dynamic_cast<PredicateLeaf&>(conditions);
+      ValueType l_vt = resolve_operand_type(leaf.get_lchild(), schema);
+      ValueType r_vt = resolve_operand_type(leaf.get_rchild(), schema);
+      if (l_vt == ValueType::VT_INVALID || r_vt == ValueType::VT_INVALID) {
+        return false;
+      }
+      if (l_vt != r_vt) {
```

Comment:
We need more check for type checking

Reply:
special check for `attr op val`, when the attr is date and val is string, try to cast it into date type.